config/crds/sync_v1alpha1_config.yaml

```diff
@@ -0,0 +1,52 @@
+apiVersion: apiextensions.k8s.io/v1beta1
+kind: CustomResourceDefinition
+metadata:
+  creationTimestamp: null
+  labels:
+    controller-tools.k8s.io: "1.0"
+  name: configs.sync.gatekeeper.sh
+spec:
+  group: sync.gatekeeper.sh
+  names:
+    kind: Config
+    plural: configs
+  scope: Namespaced
+  validation:
+    openAPIV3Schema:
+      properties:
+        apiVersion:
+          description: 'APIVersion defines the versioned schema of this representation
+            of an object. Servers should convert recognized schemas to the latest
+            internal value, and may reject unrecognized values. More info: https://git.k8s.io/community/contributors/devel/api-conventions.md#resources'
+          type: string
+        kind:
+          description: 'Kind is a string value representing the REST resource this
+            object represents. Servers may infer this from the endpoint the client
+            submits requests to. Cannot be updated. In CamelCase. More info: https://git.k8s.io/community/contributors/devel/api-conventions.md#types-kinds'
+          type: string
+        metadata:
+          type: object
+        spec:
+          properties:
+            whitelist:
```

Comment:
The idea here is "these are the allowed gvks to sync". This is related to your general comment on the commit asking why we're using a config object to govern syncing.

The answer dates to the meeting two weeks ago, where we discussed how we should decide what to sync. We had mentioned the possibility of scraping ConstraintTemplates and/or Constraints to derive a sync list, but I mentioned the possibility of using that capability to do things like access secrets beyond what RBAC allows by creating a ConstraintTemplate that has a dependency on the sensitive object and dumps its contents. IIRC the solution was to initially rely on a config so people know/control what they are replicating into OPA.

In the path where we move forward with the static analysis, this list would become the exclusive list of resources the system is permitted to sync. There would be an accompanying list that would describe resources that should not be synced. This current iteration is in anticipation of that future.

Do you agree with the above general approach? Given this context for the field, what would be a good, descriptive name?

Reply:
> Do you agree with the above general approach? Given this context for the field, what would be a good, descriptive name?

I'm curious what others think about this. @teq0 suggested having constraint templates declare their dependencies. E.g., in addition to the OpenAPI spec, `target`, and `rego`, there could also be a `dependencies` field that lists GVKs of resources to replicate. That seemed like a good first-cut to me--because you can just install templates and the system will figure it out. With the "separate CR" approach, users need to remember to check & update the Config object each time they add a template.

I'm not sure how important the whitelist support is. I'd expect organizations interested in this kind of thing would have change management processes around loading constraint templates into the cluster.

Thoughts?

Reply:
Generally I prefer to do the safest thing first in this case.

I had thought @teq0 had agreed with his "no magic" comment. Maybe I was mistaken?

Adding a separate field is a no-go I think, since the ConstraintTemplate resource is meant to be platform-neutral. Annotations would be a likely candidate for announcing dependencies.

If introspection is an option, that may be handier, if harder to implement.

The barrier to writing the announce-and collate stuff is non-trivial, and if we write introspection, we'd essentially be spending that time on a vestigial feature. From a "get an mvp done" point of view this is the most efficient.

Reply:
> I had thought @teq0 had agreed with his "no magic" comment. Maybe I was mistaken?

I've updated the meeting notes with links to the recordings. @teq0 suggested putting the dependencies on the constraint template.

> Adding a separate field is a no-go I think, since the ConstraintTemplate resource is meant to be platform-neutral. Annotations would be a likely candidate for announcing dependencies.

Is this just a problem due to static typing and Go or is it something more fundamental? We have a type discriminator on the constraint template (`target`) so having a dynamically typed `dependencies` field seems fair game IMO.

> The barrier to writing the announce-and collate stuff is non-trivial, and if we write introspection, we'd essentially be spending that time on a vestigial feature. From a "get an mvp done" point of view this is the most efficient.

We'd only be duplicating effort on the implementation that extracts dependencies from the constraint template. I'm guessing the work required to extract explicitly declared dependencies is minimal.

I'm raising this question because we've seen the usability challenge first hand with users needing to configure kube-mgmt to replicate resources into OPA. A possible middle ground would be explicitly declared dependencies on constraint templates combined with a deny list (which we could ship and default to contain secrets.)

Reply:
> > I had thought @teq0 had agreed with his "no magic" comment. Maybe I was mistaken?
> 
> I've updated the meeting notes with links to the recordings. @teq0 suggested putting the dependencies on the constraint template.

Thanks for adding the links! Here is the part of the meeting where we decided on this path:

```
 Max
35:26
Yeah, why don't I, why don't I do the the whitelist thing for now and use that as a source of truth for application and because because I think that's like a security first
35:37
kind of approach and then we can we can move on to make it a little more intelligent, I only replicating stuff policies need afterward.
 Unknown Speaker
35:51
Sounds good.
```

> 
> > Adding a separate field is a no-go I think, since the ConstraintTemplate resource is meant to be platform-neutral. Annotations would be a likely candidate for announcing dependencies.
> 
> Is this just a problem due to static typing and Go or is it something more fundamental? We have a type discriminator on the constraint template (`target`) so having a dynamically typed `dependencies` field seems fair game IMO.

I thought about this on my way home. The `Target` spec might be a good place for this. There is also the question of dependencies on foreign data and if/how we want to indicate that.

> 
> > The barrier to writing the announce-and collate stuff is non-trivial, and if we write introspection, we'd essentially be spending that time on a vestigial feature. From a "get an mvp done" point of view this is the most efficient.
> 
> We'd only be duplicating effort on the implementation that extracts dependencies from the constraint template. I'm guessing the work required to extract explicitly declared dependencies is minimal.
> 
> I'm raising this question because we've seen the usability challenge first hand with users needing to configure kube-mgmt to replicate resources into OPA. A possible middle ground would be explicitly declared dependencies on constraint templates combined with a deny list (which we could ship and default to contain secrets.)

Sounds like a worthwhile feature. Where does its priority lie in relation to the MVP?

Reply:
Another thought, we should probably not have this CRD be a "sync" config, but a general config.

Reply:
I think I've changed things enough w/ the commit I just posted that I've left us room to maneuver however we decide to proceed.

LMK if this is okay to merge.